### PR TITLE
Chain the backup delta VHD files

### PR DESCRIPTION
**Pull request by nraynaud**
The goal is for a tool like vhdimount to be able to mount any file and use it as a disk to recover specific file in it.
I forked struct-fu to add support for utf16be, but the original author is not thrilled: https://github.com/natevw/struct-fu/pull/6 There might be a decision to be made there before merging.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -93,7 +93,7 @@
     "semver": "^5.1.0",
     "serve-static": "^1.9.2",
     "stack-chain": "^1.3.3",
-    "struct-fu": "^1.0.0",
+    "@nraynaud/struct-fu": "^1.0.1",
```
Comment (julien-f) [inline package.json]: Would you mind also adding me as a GitHub/npm maintainer (increase the bus factor :) ).

File: src/xo-mixins/backups.js [hunk 2]
```diff
@@ -291,6 +291,14 @@ export default class {
     return backups.slice(i)
   }
 
+  // fix the parent UUID and filename in delta files after download from xapi or backup compression
+  async _chainDeltaVdiBackups ({handler, dir}) {
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: Could not this be directly integrated with the merge process?

File: src/xo-mixins/backups.js [hunk 3]
```diff
@@ -291,6 +291,14 @@ export default class {
...
+    const backups = await this._listVdiBackups(handler, dir)
+    for (let i = 1; i < backups.length; i++) {
+      await chainVhd(handler, dir + '/' + backups[i - 1], handler, dir + '/' + backups[i])
+    }
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: Is it necessary to do this for all entries?
Is it not done for each delta everytime it's created?

(Real questions, I may be completely wrong.)

Comment (nraynaud) [inline src/xo-mixins/backups.js]: There are 2 cases that need adjusting: the top of the delta stack when a new backup as been downloaded, and the first delta after a merge.
There is also the modification of the backup stacks saved before this change was made.

File: src/xo-mixins/backups.js (hunk 2 above)
Comment (nraynaud) [inline src/xo-mixins/backups.js]: A new backup delta will probably not be merged, and after a merge we need to change the first non-merged delta. So we are impacting delta that won't be touched by the merge.

File: package.json (hunk 1 above)
Comment (nraynaud) [inline package.json]: I will, but wouldn't you prefer forking it in the Vates organization?

Comment (olivierlambert) [inline package.json]: That would be better

Comment (olivierlambert) [inline package.json]: @nraynaud https://github.com/vatesfr/struct-fu

File: src/xo-mixins/backups.js (hunk 3 above)
Comment (julien-f) [inline src/xo-mixins/backups.js]: Ok, I agree it's far simpler to iterate over all deltas.

**Pull request merged**